### DISCUSSION TITLE
need help with tun device network setup, not tap

### DISCUSSION THREAD

**Discussion by calvin2021y** [firecracker-microvm/firecracker#4350] [Q&A]
The document said: ` Currently firecracker supports only TUN/TAP network backend with no multi queue support.`

The example only provide tap example,  and I can not find tun device example.   

I need to use tun device(layer 3),  not tap (layer 2).    

any suggestions?

**Comment by bchalios** [comment]
Hi @calvin2021y. Thanks for getting in touch. Firecracker does not support `tun` devices, at the moment. We only handle L2 devices.

**Comment by calvin2021y** [comment]
is tun device doable ?

**Comment by bchalios** [comment]
`virtio-net` is a L2 device, hence we tap instead of tun. I guess we _could_ do tun but this would need us to process the frame in Firecracker before passing it over to the tun device. I would expect that to be slower than what the kernel is able to do. Would you mind me asking what is your use-case, in other words, why tap device doesn't work for you?
